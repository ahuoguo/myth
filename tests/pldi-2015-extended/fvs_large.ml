type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type binop =
  | Add
  | Sub
  | Mul
  | Div

type exp =
  | Unit
  | BVar of nat
  | FVar of nat
  | Lam of nat * exp
  | App of exp * exp
  | Pair of exp * exp
  | Fst of exp
  | Snd of exp
  | Inl of exp
  | Inr of exp
  | Match of exp * nat * exp * nat * exp
  | Const of nat
  | Binop of exp * binop * exp

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (x, l1p) -> Cons (x, append l1p l2)
;;

let fvs_large : exp -> list |>
  { Unit => []
  | BVar (0) => [0]
  | BVar (1) => [1]
  | BVar (2) => [2]
  | FVar (0) => []
  | Lam (0, Unit) => []
  | Lam (0, BVar (1)) => [1]
  | App (Unit, Unit) => []
  | App (BVar (0), Unit) => [0]
  | App (Unit, BVar (1)) => [1]
  | App (BVar (0), BVar (1)) => [0; 1]
  | Fst (Unit) => []
  | Fst (BVar (1)) => [1]
  | Snd (Unit) => []
  | Snd (BVar (1)) => [1]
  | Pair (Unit, Unit) => []
  | Pair (BVar (0), Unit) => [0]
  | Pair (Unit, BVar (1)) => [1]
  | Pair (BVar (0), BVar (1)) => [0; 1]
  | Inl (Unit) => []
  | Inl (BVar (1)) => [1]
  | Inr (Unit) => []
  | Inr (BVar (1)) => [1]
  | Match (Unit, 0, Unit, 1, Unit) => []
  | Match (BVar (0), 0, Unit, 1, Unit) => [0]
  | Match (Unit, 0, BVar (1), 1, Unit) => [1]
  | Match (Unit, 0, Unit, 1, BVar (2)) => [2]
  | Match (BVar (2), 0, BVar (1), 1, BVar (0)) => [2; 1; 0]
  | Const (0) => []
  | Binop (BVar (0), Add, Unit) => [0]
  | Binop (Unit, Add, BVar (1)) => [1]
  } = ?
