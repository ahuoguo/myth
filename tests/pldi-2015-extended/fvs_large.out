(* Generated Thu Nov 13 01:56:25 2014 *)
let fvs_large : exp -> list =
  let rec f1 (e1:exp) : list =
    match e1 with
      | Unit -> Nil
      | BVar (n1) -> Cons (n1, Nil)
      | FVar (n1) -> Nil
      | Lam (n1, e2) -> f1 e2
      | App (e2, e3) -> append (f1 e2) (f1 e3)
      | Pair (e2, e3) -> append (f1 e2) (f1 e3)
      | Fst (e2) -> f1 e2
      | Snd (e2) -> f1 e2
      | Inl (e2) -> f1 e2
      | Inr (e2) -> f1 e2
      | Match (e2, n1, e3, n2, e4) -> (match f1 e2 with
                                         | Nil -> append (f1 e3) (f1 e4)
                                         | Cons (n3, l1) -> Cons (n3,
                                                              append (f1 e3)
                                                                (f1 e4)))
      | Const (n1) -> Nil
      | Binop (e2, b1, e3) -> append (f1 e2) (f1 e3)
  in
    f1
;;
